File: learning_go/ch05/exer05/main.exs
defmodule MathFunc do
  def prefixer(str) do
    &(str <> " " <> &1)
  end

  def divi(_a, b) when b == 0, do: {0, "division by zero"}
  def divi(a, b), do: {a / b, :ok}
end

defmodule Main do
  import MathFunc, only: [prefixer: 1, divi: 2]

  def run(args) do
    if length(args) != 1 do
      IO.puts("Usage: elixir main.exs <filename>")
      exit(:normal)
    end

    filename = List.first(args)

    count = File.read!(filename) |> String.length()
    IO.puts("File #{filename} has #{count} characters.")

    hello_prefix = prefixer("Hello")
    hello_prefix.("Bob") |> IO.puts()
    hello_prefix.("Maria") |> IO.puts()
    divi(3, 0) |> IO.inspect()
  end
end

Main.run(System.argv())

File: learning_go/ch05/exer05/main.go
package main

import (
	"errors"
	"fmt"
	"io"
	"log"
	"os"
)

func main() {
	// using prefixer function
	helloPrefix := prefixer("Hello")
	fmt.Println(helloPrefix("Bob"))   // should print Hello Bob
	fmt.Println(helloPrefix("Maria")) // should print Hello Maria
	// get a filename to read it's content
	args := os.Args
	if len(args) < 2 {
		log.Println("Usage: go run main.go <filename>")
		return
	}

	count, err := fileLen(args[1])
	if err != nil {
		log.Println("Error:", err)
	} else {
		fmt.Println("File length:", count)
	}

	result, err := div(10, 0)
	if err != nil {
		log.Println("Error:", err)
	} else {
		fmt.Println("Result:", result)
	}
}

func div(a, b int) (int, error) {
	if b == 0 {
		return 0, errors.New("division by zero")
	}
	return a / b, nil
}

func fileLen(filename string) (int, error) {
	file, err := os.Open(filename)
	if err != nil {
		log.Println(file, ": No such file or directory")
	}
	defer file.Close()
	data := make([]byte, 64)
	total := 0
	for {
		count, err := file.Read(data)
		if err != nil {
			if err == io.EOF {
				return total + count, nil
			}
			return 0, err
		}
		total += count
	}
}

func prefixer(str string) func(string) string {
	return func(s string) string {
		return str + " " + s
	}
}
